fapolicy_analyzer/ui/reducers/profiler_reducer.py

```diff
@@ -13,59 +13,145 @@
+    START_PROFILING_REQUEST,
+    START_PROFILING_RESPONSE,
+    PROFILING_KILL_RESPONSE,
+    PROFILING_DONE_EVENT,
+    PROFILING_TICK_EVENT,
+    PROFILING_EXEC_EVENT,
 )
 
 
 class ProfilerState(NamedTuple):
-    entry: Dict[str, str]
-    output: str
-    file: str
+    cmd: Optional[str]
+    pwd: Optional[str]
+    env: Optional[str]
+    uid: Optional[str]
+    pid: Optional[int]
+    duration: Optional[int]
+    running: bool
+    killing: bool
+    events_log: Optional[str]
+    stdout_log: Optional[str]
+    stderr_log: Optional[str]
+
+#
+# About these state subtypes:
+# I am using these subclasses to tag the event type, which allows
+# subscribers to route handling by type of the state received.
+# could also achieve this by adding the action type string to the
+# state but wanted to see what this looks like and how it holds up.
+# I wanted to also separate the properties into proper contexts by type
+# but that proves troublesome on some transitions, you still end up
+# having to include properties that are not applicable, just for the
+# next transition. Originally there were subtypes for each possible
+# state, but currently there are only subtypes for the states that need
+# special handling, all other cases use the base class ProfilerState.
+#
+# In summary, this approach wanted to behave like a State Machine, but didnt quite succeed.
+#
+
+
+class ProfilerStarted(ProfilerState):
+    pass
+
+
+class ProfilerTick(ProfilerState):
+    pass
+
+
+class ProfilerKill(ProfilerState):
+    pass
+
+
+class ProfilerDone(ProfilerState):
+    pass
 
 
-default_entry = {"executeText": "",
-                 "argText": "",
-                 "userText": "",
-                 "dirText": "",
-                 "envText": ""
-                 }
+def empty_profiler_state():
+    return ProfilerState(cmd=None,
+                         pwd=None,
+                         env=None,
+                         uid=None,
+                         pid=None,
+                         duration=None,
+                         running=False,
+                         killing=False,
+                         events_log=None,
+                         stdout_log=None,
+                         stderr_log=None, )
 
 
-def _create_state(state: ProfilerState, **kwargs: Optional[Any]) -> ProfilerState:
-    return ProfilerState(**{**state._asdict(), **kwargs})
+def derive_profiler_state(target, source: ProfilerState, **kwargs: Optional[Any]) -> ProfilerState:
+    return target(**{**source._asdict(), **kwargs})
 
 
-def handle_set_profiler_state(state: ProfilerState, action: Action) -> ProfilerState:
-    return _create_state(state, entry={**action.payload})
+def profiler_state(target, **kwargs: Optional[Any]) -> ProfilerState:
+    return target(**{**empty_profiler_state()._asdict(), **kwargs})
+
+
+def handle_profiler_init_state(state: ProfilerState, action: Action) -> ProfilerState:
+    return derive_profiler_state(ProfilerState, state)
```

Comment:
I don't think you need this either. This action would only come from the init of the `profiler_feature` so it would always be the default state correct?

Reply:
Correct, again :+1: